ui/components/multichain/asset-picker-amount/asset-balance/asset-balance-text.tsx

```diff
@@ -0,0 +1,125 @@
+import { useTokenTracker } from '../../../../hooks/useTokenTracker';
+import { useCurrencyDisplay } from '../../../../hooks/useCurrencyDisplay';
+import { useTokenFiatAmount } from '../../../../hooks/useTokenFiatAmount';
+import useIsFiatPrimary from '../hooks/useIsFiatPrimary';
+
+export interface AssetBalanceTextProps {
+  asset: Asset;
+  balanceColor: TextColor;
+}
+
+export function AssetBalanceText({
+  asset,
+  balanceColor,
+}: AssetBalanceTextProps) {
+  const secondaryCurrency = useSelector(getCurrentCurrency);
+
+  const isFiatPrimary = useIsFiatPrimary();
+
+  const { tokensWithBalances } = useTokenTracker({
+    tokens: [{ address: asset.details?.address }],
+    address: undefined,
+  });
+  const balanceString = (tokensWithBalances[0] as any)?.string;
+
+  const balanceValue = useSelector(getSelectedAccountCachedBalance);
+
+  const nativeTokenFiatBalance = useCurrencyDisplay(balanceValue, {
+    numberOfDecimals: 2,
+    currency: secondaryCurrency,
+  })[1].value;
+
+  const erc20TokenFiatBalance = useTokenFiatAmount(
+    asset.details?.address,
+    balanceString,
+    undefined,
+    undefined,
+    true,
+  );
+
+  const formattedFiat =
+    asset.type === AssetType.native
+      ? nativeTokenFiatBalance
+      : erc20TokenFiatBalance;
+
+  if (isFiatPrimary) {
+    return (
+      <CurrencyDisplay
+        hideLabel
+        className="currency-input__conversion-component"
+        currency={secondaryCurrency}
+        value={undefined}
+        numberOfDecimals={2}
+        displayValue={formattedFiat}
+        data-testid={undefined}
+        style={undefined}
+        prefix={undefined}
+        prefixComponent={undefined}
+        hideTitle={undefined}
+        denomination={undefined}
+        suffix={undefined}
+      />
+    );
+  } else if (asset.type === AssetType.native) {
+    return (
+      <UserPreferencedCurrencyDisplay
+        hideLabel
+        value={asset.balance}
+        type={PRIMARY}
+        textProps={{
+          color: balanceColor,
+          variant: TextVariant.bodySm,
+        }}
+        suffixProps={{
+          color: balanceColor,
+          variant: TextVariant.bodySm,
+        }}
+        data-testid={undefined}
+        ethNumberOfDecimals={undefined}
+        fiatNumberOfDecimals={undefined}
+        numberOfDecimals={undefined}
+        showEthLogo={undefined}
+        showFiat={undefined}
+        showNative={undefined}
+        showCurrencySuffix={undefined}
+      />
+    );
+  }
+  return (
+    <TokenBalance
+      hideLabel
+      token={{
+        ...asset.details,
+        decimals: asset.details?.decimals
+          ? Number(asset.details.decimals)
+          : undefined,
+      }}
+      textProps={{
+        color: balanceColor,
+        variant: TextVariant.bodySm,
+      }}
+      suffixProps={{
+        color: balanceColor,
+        variant: TextVariant.bodySm,
+      }}
+      className={undefined}
+      showFiat={undefined}
+    />
```

Comment:
those are common props btw the 3 cases of your condition statement:
```
  const commonProps = {
    hideLabel: true,
    textProps: { color: balanceColor, variant: TextVariant.bodySm },
    suffixProps: { color: balanceColor, variant: TextVariant.bodySm },
  };
```
you can use this to reduce redundancy and improve readability like this:
```
 if (isFiatPrimary) {
    return <CurrencyDisplay {...commonProps} currency={secondaryCurrency} displayValue={formattedFiat} />;
  } else if (asset.type === AssetType.native) {
    return <UserPreferencedCurrencyDisplay {...commonProps} value={asset.balance} type={PRIMARY} />;
  }

  return (
    <TokenBalance
      {...commonProps}
      token={{ ...asset.details, decimals: asset.details?.decimals ? Number(asset.details.decimals) : undefined }}
    />
  );
```

Reply:
Great call on the common props.

To avoid the undefined props, we would need to exclude that line from eslint rules, which I like to avoid. I went ahead and added typing for those components instead in ae279f141dd3c218d2d752aa5be242381848289d